aten/src/ATen/core/ivalue_inl.h

```diff
@@ -382,7 +382,13 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
         "Attempting to mark a completed Future as complete again. Note that "
         "a Future can only be marked completed once.");
 
-    preMarkCompletedHook(value, std::move(data_ptrs));
+    try {
+      preMarkCompletedHook(value, std::move(data_ptrs));
+    } catch (const std::exception&) {
+      setErrorInternal(std::current_exception(), lock);
```

Comment:
This is clearly better then the current version, but does this mean chained Futures will never be marked as completed, even if they are not going to consume the value from the root future?

Reply:
Are you talking about child futures in the sense of those returned by a call to `parent_future.then(cb)`? In that case, my understanding is that if the `preMarkCompletedHook` fails on `parent_future`, we call `setErrorInternal` on `parent_future`, which should still invoke all its callbacks, including the one for `child_future`. Those callbacks receive `parent_future` as argument, even if `parent_future` is in an error state, thus if those callbacks try to access `parent_future`'s value an exception will be thrown, which will cause `child_future` to also be set to error. If the callback never accesses `parent_future`'s value then it will just return a value as normal and `child_future` will be set correctly.

Reply:
BTW I'm adding a test for this in #56518.